indexing-hadoop/src/main/java/io/druid/indexer/IndexGeneratorJob.java

```diff
@@ -548,9 +595,29 @@ protected void reduce(
             toMerge.add(file);
 
             context.progress();
-            persist(index, interval, file, progressIndicator);
-            // close this index and make a new one, reusing same buffer
-            index.close();
+            final IncrementalIndex persistIndex = index;
+            persistFutures.add(
+                persistExecutor.submit(
+                    new ThreadRenamingRunnable(String.format("%s-persist", file.getName()))
+                    {
+                      @Override
+                      public void doRun()
+                      {
+                        try {
+                          persist(persistIndex, interval, file, progressIndicator);
+                        }
+                        catch (Exception e) {
+                          log.error("persist index error", e);
+                          throw Throwables.propagate(e);
+                        }
+                        finally {
+                          // close this index
+                          persistIndex.close();
```

Comment:
After the runnable exits, does the future still hold a non-weak reference to persistIndex?